private/buf/cmd/buf/command/beta/studioagent/studioagent.go

```diff
@@ -0,0 +1,165 @@
+		Short: "Run an HTTP server as the studio agent with the origin be set as the allowed origin for CORS options.",
+		Args:  cobra.ExactArgs(1),
+		Run: builder.NewRunFunc(
+			func(ctx context.Context, container appflag.Container) error {
+				return run(ctx, container, flags)
+			},
+			bufcli.NewErrorInterceptor(),
+		),
+		BindFlags: flags.Bind,
+	}
+}
+
+type flags struct {
+	Port              string
+	DisallowedHeaders []string
+	ForwardHeaders    []string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.Port,
+		portFlagName,
+		"8080",
+		"The port to be exposed to accept HTTP request.",
+	)
+	flagSet.StringSliceVar(
+		&f.DisallowedHeaders,
+		disallowedHeadersFlagName,
+		nil,
+		`The headers to be disallowed via the agent to the target server. Multiple headers are appended if specified multiple times.`,
+	)
+	flagSet.StringSliceVar(
+		&f.ForwardHeaders,
+		forwardHeadersFlagName,
+		nil,
+		`The headers to be forwarded via the agent to the target server. Must be a colon-separated key-value pair (like --forward-header=fromHeader1:toHeader1). Multiple header pairs are appended if specified multiple times.`,
+	)
+}
+
+func run(
+	ctx context.Context,
+	container appflag.Container,
+	flags *flags,
+) error {
+	logger := container.Logger().Named("studio-agent")
+	// convert the disallowedHeaders from a list of string to a map
+	disallowedHeaders := make(map[string]struct{}, len(flags.DisallowedHeaders))
+	for _, header := range flags.DisallowedHeaders {
+		disallowedHeaders[header] = struct{}{}
+	}
+	// convert the forwardHeaders from a list of colon-separated key-value pair string to a map
+	forwardHeaders := make(map[string]string, len(flags.ForwardHeaders))
+	for _, pair := range flags.ForwardHeaders {
+		s := strings.Split(pair, ":")
+		if len(s) != 2 {
+			return fmt.Errorf("unknown key-pair value of forward-headers: %s", pair)
+		}
+		forwardHeaders[s[0]] = s[1]
+	}
+	config, err := bufcli.NewConfig(container)
+	if err != nil {
+		return err
+	}
+	mux := bufstudioagent.NewHandler(
+		logger,
+		container.Arg(0), // the origin from command argument
+		config.TLS,
+		disallowedHeaders,
+		forwardHeaders,
+	)
+	server := http.Server{
+		Addr:    ":" + flags.Port,
+		Handler: requestLoggingHandler(mux, logger),
+	}
+	signalC, closer := interrupt.NewSignalChannel()
+	go func() {
+		logger.Info(fmt.Sprintf("listening on %s", server.Addr))
+		if err = server.ListenAndServe(); err != nil {
+			closer()
+			return
+		}
+	}()
+	<-signalC
+	if err != nil {
+		return err
+	}
+	if err := logger.Sync(); err != nil {
+		return err
+	}
+	return server.Shutdown(ctx)
+}
+
+func requestLoggingHandler(mux http.Handler, logger *zap.Logger) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		bodyBytes, err := ioutil.ReadAll(http.MaxBytesReader(w, r.Body, bufstudioagent.MaxMessageSizeBytesDefault))
```

Comment:
We shouldn't retain+log the whole body here, up to 5mb log lines wouldn't exactly be a summary. The bodies are also base64 encoded binary proto so not exactly user friendly. Depending on https://github.com/bufbuild/buf/pull/1137/files#r881775141, we may get some basic functionality to logger embedded there. 

However, logging at this level only gives us insight into the envelope whereas I think the log would be more interesting if the target url, upstream response (grpc) status would be in there somewhere. I'm not sure if there's an easy way to attach details to logs made in a wrapping handler, lets chat about this offline.

Reply:
resolve as requesting logging is removed from the PR